### PR TITLE
Add a listError option for parsing

### PR DESCRIPTION
**Pull request by gmetais**
Pull request for #59.

### CODE REVIEW

**Comment by lydell** [comment]
Why not make the `silent` option work like `listError` instead?

**Comment by gmetais** [comment]
I was afraid it wouldn't be exactly silent. Like a little noise :)

I'll change it right now.

File: lib/parse/index.js [hunk 1]
```diff
@@ -56,18 +56,31 @@ module.exports = function(css, options){
...
+    var errorObj;
+    var fullErrorMsg = options.source + ':' + lineno + ':' + column + ': ' + msg;
+
     if (options.silent === true) {
-      return false;
+      errorObj = {
+        message: fullErrorMsg
+      };
+    } else {
+      errorObj = new Error(fullErrorMsg);
```
Comment (lydell) [inline lib/parse/index.js]: Why not create real `Error` objects for the `errorsList` as well?

**Comment by lydell** [comment]
Add documentation about the new property in https://github.com/reworkcss/css#stylesheet as well.

File: lib/parse/index.js (hunk 1 above)
Comment (gmetais) [inline lib/parse/index.js]: It wouldn't stringify correctly, as the `message` property of Error is non enumerable.

```js
var err = new Error('a message');
console.log(JSON.stringify(err));
// {}
```

**Comment by gmetais** [comment]
Documentation updated :)

File: lib/parse/index.js (hunk 1 above)
Comment (lydell) [inline lib/parse/index.js]: Does that matter?

File: Readme.md [hunk 2]
```diff
@@ -114,6 +114,14 @@ The root node returned by `css.parse`.
...
   - rules: `Array` of nodes with the types `rule`, `comment` and any of the
     at-rule types.
+  - parsingErrors: `Array` of `Object`s. Errors collected during parsing when
+    option `silent` is true.
+    - `message`: `String`. The full error message with the source position.
+    - `reason`: `String`. The error message without position.
+    - `filename`: `String`.
+    - `line`: `Integer`.
+    - `column`: `Integer`.
+    - `source`: `String`. The portion of code.
```
Comment (lydell) [inline Readme.md]: This belongs better in the “Errors” section above. Just keep line 117 and 118 here.

File: lib/parse/index.js (hunk 1 above)
Comment (gmetais) [inline lib/parse/index.js]: Hmmm... ok lets do it!

File: lib/parse/index.js [hunk 3]
```diff
@@ -56,18 +56,21 @@ module.exports = function(css, options){
...
+      throw err;
     }
-
-    var err = new Error(options.source + ':' + lineno + ':' + column + ': ' + msg);
-    err.reason = msg;
-    err.filename = options.source;
-    err.line = lineno;
-    err.column = column;
-    err.source = css;
-    throw err;
```
Comment (lydell) [inline lib/parse/index.js]: No need to make such a large diff. Keep `err` as the variable name.

File: Readme.md [hunk 4]
```diff
@@ -114,6 +114,14 @@ The root node returned by `css.parse`.
 - stylesheet: `Object`:
   - rules: `Array` of nodes with the types `rule`, `comment` and any of the
     at-rule types.
+  - parsingErrors: `Array` of `Object`s. Errors collected during parsing when
```
Comment (lydell) [inline Readme.md]: Object -> Error

**Comment by gmetais** [comment]
Done :)

I removed the `error.position` part in the Error chapter, because this is not what is see in the code. Am I right?

File: Readme.md [hunk 5]
```diff
@@ -56,13 +56,23 @@ result.map // source map object
...
-without the position use `error.reason`.
+Errors thrown during parsing have the following properties:
+
+- `message`: `String`. The full error message with the source position.
+- `reason`: `String`. The error message without position.
+- `filename`: `String` or `undefined`. The value of `options.source` if
+  passed to `css.parse`. Otherwise `undefined`.
+- `line`: `Integer`.
+- `column`: `Integer`.
+- `source`: `String`. The portion of code that couldn't be parsed.
```
Comment (lydell) [inline Readme.md]: Looks good! Just remove `` ` ``s around the property names, because of commit 40098815.

File: Readme.md [hunk 6]
```diff
@@ -56,13 +56,23 @@ result.map // source map object
...
+
+When parsing with the `silent` option, errors are listed in the
+`parsingErrors` property of the [`stylesheed`](#stylesheet) object instead
```
Comment (lydell) [inline Readme.md]: styleshee<strong>d</strong> → styleshee<strong>t</strong>, object → node

File: lib/parse/index.js [hunk 7]
```diff
@@ -56,18 +56,21 @@ module.exports = function(css, options){
...
+  function error(msg) {
     var err = new Error(options.source + ':' + lineno + ':' + column + ': ' + msg);
     err.reason = msg;
     err.filename = options.source;
     err.line = lineno;
     err.column = column;
     err.source = css;
-    throw err;
+
+    if (options.silent === true) {
```
Comment (lydell) [inline lib/parse/index.js]: The rest of the code base just check for truthiness of boolean options. I’d suggest removing ` === true` for consistency.

**Comment by lydell** [comment]
This looks good to me now! Just change the last tiny nits, squash into one commit and we should be good to go! Any objections to this PR? @conradz?

File: Readme.md (hunk 6 above)
Comment (gmetais) [inline Readme.md]: Nice catch!

**Comment by gmetais** [comment]
Done, thanx for your remarks!

**Comment by lydell** [comment]
> squash into one commit

**Comment by gmetais** [comment]
Should be better now.

**Comment by lydell** [comment]
Yeah that’s better. However, the commit message body is really confusing if you don’t know that the commit is a squash. Perhaps better to only keep the first line of the commit message?

**Comment by gmetais** [comment]
Oops sorry, it was my first git squashing!

**Comment by gmetais** [comment]
Hey @lydell @conradz!

I'd :heart: a v2.2.0 release on npm with all the good stuff that's awaiting. Do you think it's possible?

**Comment by lydell** [comment]
Sure! I just realized that `silent: true` used to return `false`, but now it returns a truthy object instead.

- Is this a breaking change? If not we could release 2.2.0.
- Should we release 3.0.0?
- Should we deprecate `silent: true` and add `silent: 'list'`?

@conradz You seem to be the most active maintainer; what do you think?

**Comment by conradz** [comment]
The `error` function used to return false, but now it returns `undefined`, at least if I read it correctly, which is falsy. Also the `error` function is only used internally so I don't believe it is breaking. If this is correct I think v2.2.0 would be the correct version to use.

**Comment by lydell** [comment]
True, I misunderstood that. Then we should be good to go! I could make a release, but not tonight. What about you?

**Comment by conradz** [comment]
I could try to cut a release today if you're fine with that. I'm fine with the current usage and behavior of this option.

**Comment by lydell** [comment]
I’m fine with that. :+1:

**Comment by conradz** [comment]
`v2.2.0` is released with this added! :shipit:

**Comment by gmetais** [comment]
That's great!!

**Pull request merged**